src/lustre/lustreGenRefTypeImpNodes.ml

```diff
@@ -0,0 +1,143 @@
+    List.map (fun (p, id, ty, cl, _) -> (p, id, ty, cl)) inputs 
+  in
+  (* We generate two imported nodes: One for the input node's contract (w/ type info), and another 
+     for the input node's inputs/environment *)
+  Some (A.NodeDecl (span, (gen_node_id, extern, params, inputs2, outputs2, locals, node_items, contract)),
+        A.NodeDecl (span, (gen_node_id2, extern, params, inputs, locals_as_outputs @ outputs, [], node_items, Some base_contract)))
+
+let ref_type_to_contract: Ctx.tc_context -> A.lustre_type -> HString.t option -> A.declaration option
+= fun ctx ty node_id -> match ty with 
+  | RefinementType (pos, (_, id, ty), expr) as ref_type -> 
+    let span = { A.start_pos = Lib.dummy_pos; end_pos = Lib.dummy_pos } in
+    let ty_str = Lib.string_of_t A.pp_print_lustre_type ref_type |> HString.mk_hstring in
+    let gen_node_id = HString.concat2 (HString.mk_hstring (string_of_int !i)) 
+                                      ty_str in
+    i := !i + 1;
+    let is_extern = true in
+    let params = [] in 
+    let vars = AH.vars_without_node_call_ids expr in
+    let inputs = SI.diff vars (SI.singleton id) |> SI.elements in
+    let inputs = List.filter_map (fun id -> 
+      let ty = Ctx.lookup_ty ctx id |> unwrap in
+      let ty = Chk.expand_type_syn_reftype_history ctx ty |> unwrap_res in
+      let is_const = Ctx.member_val ctx id in
+      let call_params = (match node_id with 
+      | Some node_id -> Ctx.lookup_node_param_ids ctx node_id |> unwrap 
+      | None -> []
+      ) in
+      let is_global_const = is_const && not (List.mem id call_params) in
+      if is_global_const 
+      then None 
+      else Some (pos, id, ty, A.ClockTrue, is_const)
+    ) inputs in
+    let outputs = [(pos, id, ty, A.ClockTrue)] in
+    let locals = [] in 
+    (* To prevent slicing, we mark generated imported nodes as main nodes *)
+    let node_items = [A.AnnotMain(pos, true)] in 
+    (* Add assumption for each variable with a refinement type in 'expr' *)
+    let assumptions = List.filter_map (fun (_, id, _, _, _) -> 
+      let ty = Ctx.lookup_ty ctx id |> unwrap in 
+      match ty with 
+        | A.RefinementType (_, (_, id2, _), expr) -> 
+          let expr =  (AH.substitute_naive id2 (Ident(pos, id)) expr) in
+          Some (A.Assume (pos, None, false, expr))
+        | _ -> None 
+    ) inputs in 
+    (* Add guarantee for 'expr' *) 
+    let guarantee = A.Guarantee (pos, None, false, expr) in
+    let contract = Some (guarantee :: assumptions) in
+    Some (NodeDecl (span, (gen_node_id, is_extern, params, inputs, outputs, locals, node_items, contract)))
+  | _ -> None
+
+let gen_imp_nodes: Ctx.tc_context -> A.declaration list -> A.declaration list 
+= fun ctx decls -> 
+  List.fold_left (fun acc decl -> 
+    match decl with 
+    | A.TypeDecl (_, AliasType (_, _, ty))
+    | A.ConstDecl (_, FreeConst (_, _, ty))
+    | A.ConstDecl (_, TypedConst (_, _, _, ty)) -> 
+      (match (ref_type_to_contract ctx ty None) with  
```

Comment:
This is the only call to `ref_type_to_contract` and the third argument is `None`. Can we remove the third parameter?

Reply:
Addressed